thirdparty/hnswlib/hnswlib/space_cosine.h

```diff
@@ -15,14 +15,21 @@ CosineDistance(const void* pVect1, const void* pVect2, const void* qty_ptr) {
     return -1.0f * Cosine(pVect1, pVect2, qty_ptr);
 }
 
+static float
+CosineSQ8Distance(const void* pVect1, const void* pVect2, const void* qty_ptr) {
+    return -1.0f * faiss::ivec_inner_product((const int8_t*)pVect1, (const int8_t*)pVect2, *(size_t*)qty_ptr);
+}
+
 class CosineSpace : public SpaceInterface<float> {
     DISTFUNC<float> fstdistfunc_;
+    DISTFUNC<float> fstdistfunc_sq_;
```

Comment:
The introduction of an additional entity `fstdistfunc_sq_` is needed, because `SpaceInterface<>` struct does not distinguish between underlying datatypes, right?

Reply:
Yes